src/rgw/driver/rados/rgw_notify.cc

```diff
@@ -975,17 +982,40 @@ static inline bool notification_match(reservation_t& res,
 }
 
   int publish_reserve(const DoutPrefixProvider* dpp,
+		      const SiteConfig& site,
 		      EventType event_type,
 		      reservation_t& res,
 		      const RGWObjTags* req_tags)
 {
-  const RGWPubSub ps(res.store, res.user_tenant);
-  const RGWPubSub::Bucket ps_bucket(ps, res.bucket);
   rgw_pubsub_bucket_topics bucket_topics;
-  auto rc = ps_bucket.get_topics(res.dpp, bucket_topics, res.yield);
-  if (rc < 0) {
-    // failed to fetch bucket topics
-    return rc;
+  if (all_zonegroups_support(site, zone_features::notification_v2) &&
+      res.store->stat_topics_v1(res.user_tenant, res.yield, res.dpp) == -ENOENT) {
+    auto ret = 0;
+    if (!res.s) {
+      //  for non S3-request caller (e.g., lifecycle, ObjectSync), bucket attrs
+      //  are not loaded, so force to reload the bucket, that reloads the attr.
+      // for non S3-request caller, res.s is nullptr
+      ret = res.bucket->load_bucket(dpp, res.yield);
```

Comment:
@kchheda3 @yuvalif this call is showing up in lifecycle crashes (see https://tracker.ceph.com/issues/64571#note-10). is it possible that `res.bucket` is null here?

Reply:
it is possible. this ctor: https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_sal_rados.h#L736
does not initialize the bucket

Reply:
> it is possible. this ctor: https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_sal_rados.h#L736 does not initialize the bucket

but @yuvalif that ctor is for s3 requests, hence it has req_state and the code is checking `if (!res.s)`, so that ctor will not be invoked.
@cbodley looking at all the calls for notification for LC which are inside this [rgw_lc.cc](https://github.com/ceph/ceph/blob/main/src/rgw/rgw_lc.cc), does not look like anywhere bucket is nullptr.

Reply:
@cbodley looking at this [crash](https://tracker.ceph.com/issues/64571#note-8) which is non-notification one, it has the same later call trace which points to issue while reading the bucket instance info `read_bucket_instance_info`

Reply:
wasn't the en2end tracing regression in LC because of the infinite recurssion?

Reply:
that recursion bug showed up in librgw tests (https://tracker.ceph.com/issues/64543). the lifecycle crashes are still happening on main (https://tracker.ceph.com/issues/64571)

Reply:
i'm still digging into this lifecycle crash. there's some memory corruption around bucket attrs

i am still skeptical about this `load_bucket()` call, mainly because lifecycle has already loaded it [here in RGWLC::bucket_lc_process()](https://github.com/ceph/ceph/blob/56104de/src/rgw/rgw_lc.cc#L1700-L1701). i pushed a [wip-rgw-notify-no-load-bucket](https://github.com/ceph/ceph-ci/commits/wip-rgw-notify-no-load-bucket) branch that comments this call out, and have so far been unable to reproduce the crashes against it: https://pulpito.ceph.com/?suite=rgw:lifecycle

maybe the corruptions happens when we load attrs on top of existing attrs? cc @soumyakoduri

Reply:
my vstart reproducer doesn't catch the crash either with this call commented out. i ran the lifecycle expiration test 100 times in a row:

> ...
s3tests_boto3/functional/test_s3.py::test_lifecycle_expiration[99-100] PASSED                                                                          [ 99%]
s3tests_boto3/functional/test_s3.py::test_lifecycle_expiration[100-100] PASSED                                                                         [100%]

Reply:
@mattbenjamin @soumyakoduri @kchheda3 @yuvalif what should we do here? removing this `load_bucket()` block should be fine for lifecycle since it's already loaded the attrs, but might break the multisite sync notifications?

Reply:
that block seems to be responsible for the lifecycle crashes, but it's not clear exactly how. the lifecycle crashes (https://tracker.ceph.com/issues/64571) haven't shown up on squid yet because this change has not merged to squid yet (still staged in https://github.com/ceph/ceph/pull/56069 but we need to merge soon)

Reply:
before calling the publish_reserve, the caller needs to call `driver->get_notification` which passes the bucket. So in case of LC and replication we can see bucket is loaded correctly with attrs (lc already has bucket loaded correctly), we can confirm the same for replication and if its not loaded then before calling the `publish_reserve` we can load the bucket it [here](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_cr_rados.cc#L744).
the only issue is LC has multiple calls to `driver->get_notification`, just have to be confirmed that bucket is loaded correctly in all cases, else it could result in missing events for lifecycle

Reply:
@kchheda3 would you be willing to raise a pull request?

Reply:
> @kchheda3 would you be willing to raise a pull request?

yes i plan to, but i m still in process of testing whether that approach works.
just confirming all the lc events and replication events are generated correctly after removing the code to reload_bucket() from `publish_reserve`
as you said for `lc` the bucket is always loaded, so it should not be a problem
i just need to explicitly load the bucket for replication events in case if its not loaded

Reply:
> > i just need to explicitly load the bucket for replication events in case if its not loaded
> 
> at https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_cr_rados.cc#L779-L780 we initialize a RadosBucket with valid RGWBucketInfo, but without attrs. i guess we should call `dest_bucket.load_bucket()` up there?

that for the create_event, we need to do for [delete_event](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_cr_rados.cc#L947C21-L947C27) as well ?
this is brittle approach, coz in future for newer events, the caller will need to ensure the bucket has attr_loaded before calling `publish_reserve` ?
is there any existing api which can confirm if the `attrs` are already loaded and if loaded then only load ?
then we can just add that check in `publish_reserve` where are loading the bucket and this would not cause the reloading for lc events?

Reply:
> is there any existing api which can confirm if the `attrs` are already loaded and if not loaded then only load ?

all you can do is check whether `Bucket::get_attrs()` returns an empty map, but maybe the bucket actually loaded attrs but found none

it bothers me to no end that these sal::Object, Bucket, etc. objects have 'uninitialized' states like this (sorry @dang). that makes it impossible to reason about them in generic code like this. i think notification code should expect the caller to pass in a properly-initialized variable

Reply:
we should probably keep the fix in PR #56712 regardless. if calling `load_bucket()` could be avoided in this case, we shoudl avoid it

Reply:
we've got a couple more crashes from notification and LC setup together and both have different call stack but both fail in attrs decoding
this one is why decoding the bucket notification attrs during LC processing.
`0x00007f6233140a9f in raise () from /lib64/libc.so.6
#1  0x00007f6233113ece in abort () from /lib64/libc.so.6
#2  0x00007f62373d89b6 in ceph::__ceph_assert_fail(char const*, char const*, int, char const*) () from /usr/lib64/ceph/libceph-common.so.2
#3  0x00007f62373d8b22 in ceph::__ceph_assert_fail(ceph::assert_data const&) ()
   from /usr/lib64/ceph/libceph-common.so.2
#4  0x00007f62384659ba in ceph::buffer::v15_2_0::ptr::copy_out(unsigned int, unsigned int, char*) const () from /lib64/librados.so.2
#5  0x00007f623846a420 in ceph::buffer::v15_2_0::list::iterator_impl<true>::copy(unsigned int, char*) () from /lib64/librados.so.2
#6  0x00005590bfe34293 in ceph::decode(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, ceph::buffer::v15_2_0::list::iterator_impl<true>&) ()
#7  0x00005590c01131e6 in rgw_pubsub_dest::decode(ceph::buffer::v15_2_0::list::iterator_impl<true>&) ()
#8  0x00005590c01135bf in rgw_pubsub_topic::decode(ceph::buffer::v15_2_0::list::iterator_impl<true>&) ()
#9  0x00005590c011666e in rgw_pubsub_topic_filter::decode(ceph::buffer::v15_2_0::list::iterator_impl<true>&) ()
#10 0x00005590c011871f in rgw_pubsub_bucket_topics::decode(ceph::buffer::v15_2_0::list::iterator_impl<true>&) ()
#11 0x00005590c02cead8 in get_bucket_notifications(DoutPrefixProvider const*, rgw::sal::Bucket*, rgw_pubsub_bucket_topics&) ()
#12 0x00005590c00baeda in rgw::notify::publish_reserve(DoutPrefixProvider const*, std::vector<rgw::notify::EventType, std::allocator<rgw::notify::EventType> > const&, rgw::notify::reservation_t&, RGWObjTags const*) ()
#13 0x00005590c0284f3e in remove_expired_obj(DoutPrefixProvider const*, lc_op_ctx&, bool, std::vector<rgw::notify::EventType, std::allocator<rgw::notify::EventType> > const&) ()
#14 0x00005590c029eea9 in LCOpAction_CurrentExpiration::process(lc_op_ctx&) ()
#15 0x00005590c0286acf in LCOpRule::process(rgw_bucket_dir_entry&, DoutPrefixProvider const*, WorkQ*) ()
#16 0x00005590c0290063 in RGWLC::bucket_lc_process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, RGWLC::LCWorker*, long, bool)::{lambda(RGWLC::LCWorker*, WorkQ*, boost::variant<void*, std::tuple<LCOpRule, rgw_bucket_dir_entry>, std::tuple<lc_op, rgw_bucket_dir_entry>, rgw_bucket_dir_entry>&)#8}::operator()(RGWLC::LCWorker*, WorkQ*, boost::variant<void*, std::tuple<LCOpRule, rgw_bucket_dir_entry>, std::tuple<lc_op, rgw_bucket_dir_entry>, rgw_bucket_dir_entry>&) const ()
#17 0x00005590c029a0ad in WorkQ::entry() ()
#18 0x00007f62341031cf in start_thread () from /lib64/libpthread.so.0
#19 0x00007f623312bdd3 in clone () from /lib64/libc.so.6`

and this one is during get_obj_state

`#0  0x00007f09afaded6a in tc_newarray () from /lib64/libtcmalloc.so.4
#1  0x00007f09b306e755 in ceph::buffer::v15_2_0::ptr_node::cloner::operator()(ceph::buffer::v15_2_0::ptr_node const&) () from /lib64/librados.so.2
#2  0x0000561cb3d4a973 in rgw_filter_attrset(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ceph::buffer::v15_2_0::list, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ceph::buffer::v15_2_0::list, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >*) ()
#3  0x0000561cb3ef8fb2 in RGWRados::raw_obj_stat(DoutPrefixProvider const*, rgw_raw_obj&, unsigned long*, std::chrono::time_point<ceph::real_clock, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> > >*, unsigned long*, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ceph::buffer::v15_2_0::list, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >*, ceph::buffer::v15_2_0::list*, RGWObjVersionTracker*, optional_yield) ()
#4  0x0000561cb3f17e04 in RGWRados::get_obj_state_impl(DoutPrefixProvider const*, RGWObjectCtx*, RGWBucketInfo&, rgw_obj const&, RGWObjState**, RGWObjManifest**, bool, optional_yield, bool) ()
#5  0x0000561cb3f19907 in RGWRados::get_obj_state(DoutPrefixProvider const*, RGWObjectCtx*, RGWBucketInfo&, rgw_obj const&, RGWObjState**, RGWObjManifest**, bool, optional_yield, bool) ()
#6  0x0000561cb3f1a7f7 in RGWRados::Object::get_state(DoutPrefixProvider const*, RGWObjState**, RGWObjManifest**, bool, optional_yield, bool) ()
#7  0x0000561cb3f246fa in RGWRados::Object::Delete::delete_obj(optional_yield, DoutPrefixProvider const*) ()
#8  0x0000561cb3ce23d2 in rgw::sal::RadosObject::RadosDeleteOp::delete_obj(DoutPrefixProvider const*, optional_yield) ()
#9  0x0000561cb3e7df5d in remove_expired_obj(DoutPrefixProvider const*, lc_op_ctx&, bool, std::vector<rgw::notify::EventType, std::allocator<rgw::notify::EventType> > const&) ()
#10 0x0000561cb3e97ea9 in LCOpAction_CurrentExpiration::process(lc_op_ctx&) ()
#11 0x0000561cb3e7facf in LCOpRule::process(rgw_bucket_dir_entry&, DoutPrefixProvider const*, WorkQ*) ()
#12 0x0000561cb3e89063 in RGWLC::bucket_lc_process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, RGWLC::LCWorker*, long, bool)::{lambda(RGWLC::LCWorker*, WorkQ*, boost::variant<void*, std::tuple<LCOpRule, rgw_bucket_dir_entry>, std::tuple<lc_op, rgw_bucket_dir_entry>, rgw_bucket_dir_entry>&)#8}::operator()(RGWLC::LCWorker*, WorkQ*, boost::variant<void*, std::tuple<LCOpRule, rgw_bucket_dir_entry>, std::tuple<lc_op, rgw_bucket_dir_entry>, rgw_bucket_dir_entry>&) const ()
#13 0x0000561cb3e930ad in WorkQ::entry() ()`

@cbodley you think this also could be due to multiple LC worker threads updating the bucket->handle(attrs) ??

Reply:
> @cbodley you think this also could be due to multiple LC worker threads updating the bucket->handle(attrs) ??

@kchheda3 assuming you haven't applied the fix from https://github.com/ceph/ceph/pull/56712, that sounds like a plausible explanation for both

Reply:
> > @cbodley you think this also could be due to multiple LC worker threads updating the bucket->handle(attrs) ??
> 
> @kchheda3 assuming you haven't applied the fix from #56712, that sounds like a plausible explanation for both

yes i haven't applied that band-aid, but my understanding was with that band-aid the crash that it was fixing was different one `load_bucket()->read_bucket_instance_info()`.. i guess this one
`#4  0x000055dc73b9c3b0 in std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >* std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::_Reuse_or_alloc_node::operator()<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> const&>(std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> const&) ()
#5  0x000055dc73b9c628 in std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >* std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::_M_copy<false, std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::_Reuse_or_alloc_node>(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >*, std::_Rb_tree_node_base*, std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::_Reuse_or_alloc_node&) [clone .isra.0] ()
#6  0x000055dc73b9c65d in std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >* std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::_M_copy<false, std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::_Reuse_or_alloc_node>(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >*, std::_Rb_tree_node_base*, std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::_Reuse_or_alloc_node&) [clone .isra.0] ()
#7  0x000055dc73b9ca65 in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >::operator=(std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > > const&) ()
#8  0x000055dc74082ff2 in RGWSI_Bucket_SObj::read_bucket_instance_info(ptr_wrapper<RGWSI_MetaBackend::Context, 4>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RGWBucketInfo*, std::chrono::time_point<ceph::real_clock, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> > >*, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ceph::buffer::v15_2_0::list, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v15_2_0::list> > >*, optional_yield, DoutPrefixProvider const*, rgw_cache_entry_info*, boost::optional<obj_version>) ()
#9  0x000055dc7422c3cc in std::_Function_handler<int (ptr_wrapper<RGWSI_MetaBackend::Context, 4>&), RGWBucketCtl::read_bucket_instance_info(rgw_bucket const&, RGWBucketInfo*, optional_yield, DoutPrefixProvider const*, RGWBucketCtl::BucketInstance::GetParams const&)::{lambda(ptr_wrapper<RGWSI_MetaBackend::Context, 4>&)#1}>::_M_invoke(std::_Any_data const&, ptr_wrapper<RGWSI_MetaBackend::Context, 4>&) ()
#10 0x000055dc74244cb4 in std::_Function_handler<int (RGWSI_MetaBackend_Handler::Op*), RGWBucketInstanceMetadataHandler::call(std::optional<std::variant<RGWSI_MetaBackend_CtxParams_SObj> >, std::function<int (ptr_wrapper<RGWSI_MetaBackend::Context, 4>&)>)::{lambda(RGWSI_MetaBackend_Handler::Op*)#1}>::_M_invoke(std::_Any_data const&, RGWSI_MetaBackend_Handler::Op*&&) ()
#11 0x000055dc740a71de in std::_Function_handler<int (RGWSI_MetaBackend::Context*), RGWSI_MetaBackend_Handler::call(std::optional<std::variant<RGWSI_MetaBackend_CtxParams_SObj> >, std::function<int (RGWSI_MetaBackend_Handler::Op*)>)::{lambda(RGWSI_MetaBackend::Context*)#1}>::_M_invoke(std::_Any_data const&, RGWSI_MetaBackend::Context*&&) ()
#12 0x000055dc740a897f in RGWSI_MetaBackend_SObj::call(std::optional<std::variant<RGWSI_MetaBackend_CtxParams_SObj> >, std::function<int (RGWSI_MetaBackend::Context*)>) ()
#13 0x000055dc740a7517 in RGWSI_MetaBackend_Handler::call(std::optional<std::variant<RGWSI_MetaBackend_CtxParams_SObj> >, std::function<int (RGWSI_MetaBackend_Handler::Op*)>) ()
#14 0x000055dc742317ab in RGWBucketCtl::read_bucket_instance_info(rgw_bucket const&, RGWBucketInfo*, optional_yield, DoutPrefixProvider const*, RGWBucketCtl::BucketInstance::GetParams const&) ()
#15 0x000055dc73f6ead5 in rgw::sal::RadosBucket::load_bucket(DoutPrefixProvider const*, optional_yield, bool) ()
#16 0x000055dc73f43bdd in rgw::notify::publish_reserve(DoutPrefixProvider const*, std::vector<rgw::notify::EventType, std::allocator<rgw::notify::EventType> > const&, rgw::notify::reservation_t&, RGWObjTags const*) ()
#17 0x000055dc7410cf3e in remove_expired_obj(DoutPrefixProvider const*, lc_op_ctx&, bool, std::vector<rgw::notify::EventType, std::allocator<rgw::notify::EventType> > const&) ()
#18 0x000055dc74126ea9 in LCOpAction_CurrentExpiration::process(lc_op_ctx&) ()
#19 0x000055dc7410eacf in LCOpRule::process(rgw_bucket_dir_entry&, DoutPrefixProvider const*, WorkQ*) ()
#20 0x000055dc74118063 in RGWLC::bucket_lc_process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, RGWLC::LCWorker*, long, bool)::{lambda(RGWLC::LCWorker*, WorkQ*, boost::variant<void*, std::tuple<LCOpRule, rgw_bucket_dir_entry>, std::tuple<lc_op, rgw_bucket_dir_entry>, rgw_bucket_dir_entry>&)#8}::operator()(RGWLC::LCWorker*, WorkQ*, boost::variant<void*, std::tuple<LCOpRule, rgw_bucket_dir_entry>, std::tuple<lc_op, rgw_bucket_dir_entry>, rgw_bucket_dir_entry>&) const ()`